### ISSUE TITLE
NFS: World X,Y,Z structs (Also Player->Speed is incorrect)

### ISSUE THREAD

**Issue by berkay2578** [DennisStanistan/NFSScript#13]
Hey, I'm here to make you hate creating this project again. (Because you have to deal with me, get it?)

Firstly, Speed offset(0x28) is wrong. 0x28 is Y-oriented velocity and I forgot to remove Speed after adding the new values. Change it to the pointer below please, but remember it is read only.

So anyways, I just found out that World does things a little different with X, Y and Z values. An example of this is `CDActionDebug`'s rotation values (It calculates Z-oriented rotation with two values, one for -∞ and one for +∞. Look at here to see it: nfsw.exe+3DBCC6, the 6th variable [ebp-0x18]). Game mathematically creates a definitive value through two inversely proportional values, you can see this also in speed calculation:
```
// This struct is X, Z and then Y; not X, Y and Z. Coordinates struct is Y, Z and X. Because reasons.
float VELOCITY_TOWARDS_SOUTH = objectBase + 0x20; // Towards +X
float VERTICAL_VELOCITY      = objectBase + 0x24; // Towards +Z
float VELOCITY_TOWARDS_EAST  = objectBase + 0x28; // Towards +Y
```
Through disassembling you can find that game-with usage of scalar float functions like mulss;sqrtss;subss-kind of performs triangulation(but since there are only two values this is more like a parallax) to calculate the speed in meters per second. You can take a look at this pointer to see it for yourself: (put a breakpoint to see what writes to it)
```
base: nfsw.exe+91F9D0
offset: 0x68
value: float
```
However, since I'm utter shit at math, I have no idea what the heck is going on. Maybe someone good enough with math can help, it would help World development tremendously as this is somewhat a norm for NFS:World.

**Comment by DennisStanistan** [comment]
Y, Z and X? That's quite interesting. Also this all a bit too sudden and confusing for me probably because it's 3AM as of reading this issue.

> Firstly, Speed offset(0x28) is wrong. 0x28 is Y-oriented velocity and I forgot to remove Speed after adding the new values. Change it to the pointer below please, but remember it is read only.

If it's read only then we're gonna have a problem with this function till we figure out how to change the speed:
```c#
        /// <summary>
        /// Instantly stops the <see cref="Player"/>'s car.
        /// </summary>
        public static void ForceStop()
        {
            Speed = 0f;
        }
```
This should be good for now:
```c#
        /// <summary>
        /// The <see cref="Player"/>'s car's speed in MPS.
        /// </summary>
        /// <remarks>
        /// See <see cref="Math.Mathf.ConvertSpeed(float, SpeedMeasurementConversionTypes)"/>.
        /// </remarks>
        public static float Speed
        {
            get
            {
                // TODO: Add constants to PlayerAddrs.
                int address = memory.ReadInt32((IntPtr)memory.getBaseAddress + 0x91F9D0);
                address = memory.ReadInt32((IntPtr)address + 0x68);

                return memory.ReadFloat((IntPtr)address);
            }
        }
```

**Comment by berkay2578** [comment]
Because speed is calculated through the offsets 0x20 and 0x28, setting both of them to 0 will cause the car to stop instantly. `ForceStop()` can have this inside in the meanwhile:
```
int addr = Game.PWorld_Cars + CarOffset;
memory.WriteFloat((IntPtr)(addr + (memory.getBaseAddress + GameAddrs.PSTATIC_CAR_VELOCITY_TOWARDS_SOUTH)), 0f);
memory.WriteFloat((IntPtr)(addr + (memory.getBaseAddress + GameAddrs.PSTATIC_CAR_VELOCITY_TOWARDS_EAST)), 0f);
```

**Comment by DennisStanistan** [comment]
Alright, also there is a stack overflow issue when using variables that have the following code:
```c#
int addr = Game.PWorld_Cars + CarOffset;
```
It's because of the variable `CarOffset`
```c#
        /// <summary>
        /// Gets the local player's memory offset relative to <see cref="Game.PWorld_Cars"/>.
        /// </summary>
        public static int CarOffset
        {
            get
            {
                for (int i = 0; i < 0x78; i++)
                {
                    int offset = 0xB0 * i;
                    int addr = Game.PWorld_Cars + CarOffset;

                    float localX = memory.ReadFloat((IntPtr)memory.getBaseAddress + PlayerAddrs.NON_STATIC_PLAYER_X_POS);
                    float localY = memory.ReadFloat((IntPtr)memory.getBaseAddress + PlayerAddrs.NON_STATIC_PLAYER_Y_POS);
                    float localZ = memory.ReadFloat((IntPtr)memory.getBaseAddress + PlayerAddrs.NON_STATIC_PLAYER_Z_POS);
                    float opponentX = memory.ReadFloat((IntPtr)(addr + (memory.getBaseAddress + GameAddrs.PSTATIC_CAR_X_POS)));
                    float opponentY = memory.ReadFloat((IntPtr)(addr + (memory.getBaseAddress + GameAddrs.PSTATIC_CAR_Y_POS)));
                    float opponentZ = memory.ReadFloat((IntPtr)(addr + (memory.getBaseAddress + GameAddrs.PSTATIC_CAR_Z_POS)));

                    if (Mathf.Abs(localX - opponentX) < 1f &&
                        Mathf.Abs(localY - opponentY) < 1f &&
                        Mathf.Abs(localZ - opponentZ) < 1f)
                    {
                        return offset;
                    }
                }
                return 0xFF;
            }
        }
```
And honestly, I can't figure out why it causes a stack overflow.

Interestingly enough the variable `Game.PWorld_Cars` already has `memory.getBaseAddress` so why it's added twice in VelocityTowardsX variables?
```c#
int addr = Game.PWorld_Cars + CarOffset;
return memory.ReadFloat((IntPtr)(addr + (memory.getBaseAddress + GameAddrs.PSTATIC_CAR_VELOCITY_TOWARDS_SOUTH)));
```

**Comment by berkay2578** [comment]
[My repo doesn't have that, I think you forgot that it was a pointer while _manually_ merging the pull request.](https://github.com/berkay2578/NFSScript/blob/efb242a0f6b4c55cc34f43fbde4b7d6df7ff703c/World/Player.cs)

About the stack overflow exception, it's because the loop is basically infinite without doing anything.
Just add a `-` for opponentY and it will work: `float opponentY = -memory.Read<float>(addr + GameAddrs.PSTATIC_CAR_Y_POS);`

**Comment by DennisStanistan** [comment]
I should not be manually merging pull requests, especially at night-time, whoops.. I'll fix that and attempt at doing some calculations and change the speed.

**Comment by berkay2578** [comment]
Cheers, I'll look into #10 and #11 meanwhile.

**Comment by DennisStanistan** [comment]
I changed Player.cs to yours and I still get stack overflow errors with variables that use `CarOffset`.

Are you sure that the line
```c#
int addr = Game.PWorld_Cars + CarOffset;
```
Should be in the variable `CarOffset`?

**Comment by berkay2578** [comment]
Oh, my. Yeah, that also is an issue it seems. 🙃

**Comment by DennisStanistan** [comment]
Looks like everything slowly falls apart with NFS: World
`(2017-06-08 17:35:54) EXCEPTION: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ArgumentException: Destination array is not long enough to copy all the items in the collection. Check array index and length.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.BitConverter.ToSingle(Byte[] value, Int32 startIndex)
   at NFSScript.ExtensionMethods.GetObject[T](Byte[] byteArray)
   at NFSScript.World.Player.Car.get_CarOffset()
   at NFSScript.World.Player.Car.get_VelocityTowardsEast()
   at MyMod.MyMod.OnKeyDown(Keys key)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at NFSScriptLoader.NReflec.CallMethodFromType(Type t, String methodName, Object[] o)
   at NFSScriptLoader.ModScript.CallModFunction(ModMethod modMethod, Object[] o)`

**Comment by berkay2578** [comment]
> It seems that it happens with genericMemory.
For pointer base you need `isRelativeToMemoryBase = true` and for pointer offsets you need `isRelativeToMemoryBase = false`. The stack trace basically tells 'there isn't anything at the address'.

**Comment by DennisStanistan** [comment]
Yeah, I fixed it already. Now that all this mess is out of the way, I can finally start working.

**Comment by berkay2578** [comment]
I found it, and I'm sad to say it was simple geometry instead of some advanced physics formula. Remember Pythagorean triple from high school math? Yeah...
`Speed = SquareRoot((VelocityTowardsX)^2 + (VelocityTowardsY)^2)`

In c# that's:
`double Speed = Math.Sqrt(Math.Pow(VelocityTowardsX, 2) + Math.Pow(VelocityTowardsY, 2));`
You can optimize the Power functions to:
```
double xVelocityPow2 = VelocityTowardsX * VelocityTowardsX;
double yVelocityPow2 = VelocityTowardsY * VelocityTowardsY;
double Speed = Math.Sqrt(xVelocityPow2 + yVelocityPow2);
```
You can further optimize it with using Newton's method to calculate an approximate square root, like:
```
public double NewtonSqrt(double num, int iterations = 5)
{
    double sqrt = 1;
    while (iterations > 0) {
        sqrt = (num / (2d * sqrt)) + sqrt / 2d;
        iterations--;
    }
    return sqrt;
}
```

**Comment by DennisStanistan** [comment]
> I found it, and I'm sad to say it was simple geometry instead of some advanced physics formula. Remember Pythagorean triple from high school math? Yeah...

Yeah I know, I was a head of you in that part, this is what I did before I went to work:
```c#
float east = System.Math.Pow((float)Player.Car.VelocityTowardsEast, 2);
float south = System.Math.Pow((float)Player.Car.VelocityTowardsSouth, 2);
float vertical = System.Math.Pow((float)Player.Car.VerticalVelocity, 2);

float speed = (float)System.Math.Sqrt(east + south + vertical);
```

![Image](https://content.screencast.com/users/Dennis69/folders/Jing/media/fe5a1acb-6280-4adb-b24c-8326c4ce69b1/2017-06-08_1748.png)

Just took the velocity values, put them in a calculator and hey hey hey I got the speed value in MPS.

**Comment by berkay2578** [comment]
Ok, nice, now there is only 1 problem left. How do we find out the velocity values from speed? Mathematically, it's possible and we can find a _list_ of possible values. I think that with the usage of rotation and lift values, we can sort of estimate the values. I have already semi-finished the rotation part of it on my upper comment.

**Comment by DennisStanistan** [comment]
It's possible, however I know how to do it in 2D:
```c
velX = speed*cos(angle)
velY = speed*sin(angle)
```
I'll try to figure it out and keep you updated.

**Comment by DennisStanistan** [comment]
From what I gathered so far is in order to convert speed from the velocity XYZ we need angles.

So I wrote a rotation variable
```c#
            /// <summary>
            /// The <see cref="Player"/>'s car rotation.
            /// </summary>
            public static Quaternion Rotation
            {
                get
                {
                    int addr = Game.PWorld_Cars + CarOffset;
                    float q1 = memory.ReadFloat((IntPtr)(addr + (GameAddrs.PSTATIC_CAR_ANGULAR_VELOCITY + 0x5C)));
                    float q2 = memory.ReadFloat((IntPtr)(addr + (GameAddrs.PSTATIC_CAR_ANGULAR_VELOCITY + 0x60)));
                    float q3 = memory.ReadFloat((IntPtr)(addr + (GameAddrs.PSTATIC_CAR_ANGULAR_VELOCITY + 0x64)));
                    float q4 = memory.ReadFloat((IntPtr)(addr + (GameAddrs.PSTATIC_CAR_ANGULAR_VELOCITY + 0x54)));

                    return new Quaternion(q1, q2, q3, q4);
                }
            }
```
And tried to convert it to angles but failed pretty hard, so I got stuck here.

I used an online converter to convert quaternions to angles, then I tried to do math.
I searched the web on how to convert speed to XYZ velocity but failed, so I tried to figure it out by myself.

I have these variables:
```
Speed: 55.06196213
Vertical Velocity: -0.04592379928
Quat rot: W = -0.04555625 X = 0.606118 Y = -0.07070865 Z = -0.7922252
Rotation: X = -33.024 Y = -72.166 Z = 149.037
Position: X = 3908.239 Y = -859.5852 Z = 46.60241
```

Then I just started messing around with my calculator till I get the vertical velocity value.
Tried using quat rotations, tried to take the player's X/Y position and divide it by speed, cos. tan, sin, whatever.. I tried a lot of stuff and calculations to get vertical velocity value but no success so far.

**Comment by fallk** [comment]
> I searched the entire world wide web on how to convert speed to XYZ velocity but failed, so I tried to figure it out by myself.

This is what I used in a game I made a while back
```java
double dirX = dist*(cos(toRadians(yaw))*cos(toRadians(pitch)));
double dirY = dist*(-sin(toRadians(yaw))*cos(toRadians(pitch)));
double dirZ = dist*(sin(toRadians(yaw)));
```

Of course I don't know how the coordinates in NFS are so you might need to swap the Y and Z or change the `-sin*cos` to `sin*-cos` or something like that.

Hope this helps,
CH

**Comment by DennisStanistan** [comment]
```c#
double velX = Player.Car.Speed * (Math.Cos(Mathf.DegreeToRadian(quat.x)) * Math.Cos(Mathf.DegreeToRadian(quat.y)));
double velY = Player.Car.Speed * (-Math.Sin(Mathf.DegreeToRadian(quat.x)) * Math.Cos(Mathf.DegreeToRadian(quat.y)));
double velZ = Player.Car.Speed * (Math.Cos(Mathf.DegreeToRadian(quat.x)));
```

```
(2017-06-10 06:33:54) ORIGINAL VEL: : X = -33.53981 Y = -0.6440535 Z = -28.71531
(2017-06-10 06:33:54) CALCULATED VEL: : X = -∞ Y = -∞ Z = -∞
```
Uh oh...

**Comment by fallk** [comment]
I don't think you really understood my example (I don't understand it either, don't worry)
Here's a slightly more self explanatory example:
```java
/*
From stack overflow:

x = r sin θ cos ϕ

y = r sin θ sin ϕ

z = r cos θ

Where r is your speed, ϕ is your horizontal bearing, and θ is your vertical bearing.
 */

// if you set x to 90 then it all goes +x. -90 it all goes -x.
double angX = toRadians(0); // degrees
// if you set z to 90 then it all goes +z. rinse and repeat.
double angZ = toRadians(90); // degrees
// try out how it differs as you have both at 90 or 45, for instance.

double r = 70; // speed
double x = r * sin(angX) * cos(angZ);
double y = r * sin(angX) * sin(angZ);
double z = r * cos(angX);

System.out.println("speed: " + r + "\n" +
		"dirX: " + x + " vs east: " + angX + "\n" +
		"dirY: " + y + "\n" +
		"dirZ: " + z + " vs south: " + angZ + "\n"
);
```

And here's the same thing with the full quad rotation (I haven't tested this well enough but it should work)
```java
// you know how it goes by now
double angX = toRadians(90); // degrees
double angY = toRadians(0); // degrees
double angZ = toRadians(0); // degrees

double r = 70; // speed
double x = r * sin(angX) * cos(angZ) * cos(angY);
double y = r * sin(angY) * cos(angX) * cos(angZ);
double z = r * cos(angX) * cos(angY);

System.out.println("speed: " + r + "\n" +
		"dirX: " + x + " vs east: " + angX + "\n" +
		"dirY: " + y + " vs vert: " + angY + "\n" +
		"dirZ: " + z + " vs south: " + angZ + "\n"
);
```

Any way I see you're not too keen on fussing about with this rotation garbage for much longer so I'll try and see if I can get it working, no promises though.

**Comment by DennisStanistan** [comment]
> I don't think you really understood my example (I don't understand it either, don't worry)

Yeah, I didn't understand why you had dist in there which usually stands for Distance.

> Any way I see you're not too keen on fussing about with this rotation garbage for much longer so I'll try to get it working, no promises though.

Please do, my brain is still fried from the madness occurred earlier in #10

**Comment by berkay2578** [comment]
```
float q1 = memory.ReadFloat((IntPtr)(addr + (GameAddrs.PSTATIC_CAR_ANGULAR_VELOCITY + 0x5C)));
float q2 = memory.ReadFloat((IntPtr)(addr + (GameAddrs.PSTATIC_CAR_ANGULAR_VELOCITY + 0x60)));
float q3 = memory.ReadFloat((IntPtr)(addr + (GameAddrs.PSTATIC_CAR_ANGULAR_VELOCITY + 0x64)));
float q4 = memory.ReadFloat((IntPtr)(addr + (GameAddrs.PSTATIC_CAR_ANGULAR_VELOCITY + 0x54)));
```
Some more info I've found recently:
- All direction values, even though you can change them to anything to have more visible effects, dumb down to -0.7 - 0.7.
- `CarOffset + 0x0`, the angle of the car's head: http://i.imgur.com/H8OxpWo.png
- `CarOffset + 0x8`, the angle of the chasis: http://i.imgur.com/niqK2R4.png

But, yes, I agree that finding something that directly affects the speed is much more easier.

**Comment by fallk** [comment]
> I recommend using string.Format instead of combining strings with the + operator, I know it's just an example but it's something to keep in mind,

Sorry, that example was in Java where string concatenation uses StringBuilder internally and Format uses regex so it's slower. But are you really concerned about the performance of strings in a project where you inject ASM at runtime? :P I guess with C#6 string interpolation it's a non-issue.

Anyway I've been tinkering with World this last afternoon and everything is pain. I think you should just use the yaw axis and ignore the vehicle's pitch, it should work for most cases (this is a racing game, not a flight simulator!)

**Comment by DennisStanistan** [comment]
> But are you really concerned about the performance of strings in a project where you inject ASM at runtime?

Optimizing strings makes me sleep better at night plus it wouldn't make stuff slower.

> Anyway I've been tinkering with World this last afternoon and everything is pain. I think you should just use the yaw axis and ignore the vehicle's pitch, it should work for most cases (this is a racing game, not a flight simulator!)

True, I'll keep this issue open in case something new comes up.

**Comment by berkay2578** [comment]
With the closure of this issue, I quit trying to help. I can't help someone that doesn't respect others' codes.

**Comment by fallk** [comment]
Are the two of you really gonna keep bickering about over this?

Dennis, the commit was good. The generics are fine. If you want to optimize your code, go back to ASI mods in straight C++. The nanoseconds you want to save aren't worth it. I used to be a programmer like you, now I'm sitting on 400 thousand lines of unusable code I've got to rewrite because it is completely impossible to read since I tried to optimize it as much as humanly possible. [This shit comes back to bite you.](https://en.wikipedia.org/wiki/Technical_debt)

Berkay, I still stand by you, but please be understanding of others' code conventions. You'll never be able to agree on everything. This kind of faffing about goes all the way back to the early days of Lisp. You're never gonna solve it by giving up whenever someone makes a decision you don't like.

**Comment by berkay2578** [comment]
There's a fine line between 'this is how I like to code' and just straight up narcissism. He calling the generic memory class shit publicly is not the deal here, but since we are on it: Dennis, because you are constantly adding baseaddress to intptr-casted ints in VAMemory, you are actually wasting cycles. I had uploaded a simple bench on GMemory vs VAMemory back in my VAMemory overhaul PR and GMemory was ~1 second faster because of this. Alongside this, repetition and a special function for each type actually makes c# load more stuff into memory and also makes the assembly larger in size. In the end, it's always going to run slower than GMemory.

However, this is not the deal here. What finally threw me off was his hissy fit about the `SetVariables` inside GMemory. Since I wanted to stick to VAMemory structure, I decided that the property `getBaseAddress`, which set variables and returned baseAdress, meant something and turned it into a function. He talked to length about how wrong the function was when all I did was wrap it in a function. That's called narcissism, blaming others when you're clearly on the wrong side. After the GMemory debacle I decided to cool off for a while but this is just some next-level shit. You can see and even feel that he's afraid to admit others' works being superior. Also, just to top this off, this project uses the latest c# which actually encourages `dynamic` and `T variable` in order to reduce code clutter.

I just can't man, not anymore. Enough is enough.

**Comment by DennisStanistan** [comment]
I am not sure you understand the meaning of narcissism, I am not narcissistic. Your code had various of flaws in it while they aren't MAJOR I figured that I should still point them out in a PM with you, it looked like I was REALLY pissed but in fact it was in a non-serious manner and then you snapped. Whatever.

Fallk, this has nothing to do with GMemory. The project needs to be consistent in terms of code, the project has standards that needs to be followed. You got tons of different classes that use VAMemory and then you got one class that uses GMemory, not very consistent. Berkay knows that I am not going to allow two memory classes to be used at the same time yet he still changed it to GMemory, when I confronted him about it he kept on saying shit like "dis my territory bitch" "cmon sucky sucky", if people use it for mods then that's fine but I'd rather get rid of VAMemory & GMemory and write a memory class that won't have this IntPtr conversation mess and cycle waste that Berkay claims it has.

>What's more is that he hadn't done the VAMemory, he yelled at me for keeping something he hadn't even written.

Now either that's a straight out lie or you took something I wrote out of context.

> You can see and even feel that he's afraid to admit others' works just working

In our chat in discord I eventually gave up when I saw you changed the entire of World/Game.cs to use GMemory and I asked if it at least works? You said yes and I was just ready to approve the pull request.
But then I saw this:
![Image](https://content.screencast.com/users/Dennis69/folders/Jing/media/8fb61fb6-2485-491e-bbdd-2817d55004ce/2017-06-25_1711.png)

And this
```cs
        /// <summary>
        /// 
        /// </summary>
        public long SetVariables()
        {
            this.baseAddress = (IntPtr)0;
            this.processModule = this.mainProcess[0].MainModule;
            this.baseAddress = this.processModule.BaseAddress;
            return (long)this.baseAddress;
        }
```
The function sets variables and for some reason it returns a long, why?
```cs
        /// <summary>
        /// 
        /// </summary>
        public GMemory(string pProcessName)
        {
            ProcessName = pProcessName;
            CheckProcess();
            SetVariables();
        }
```
Why does it return the base address?
```cs
this.baseAddress = (IntPtr)0;
```
This one just makes me laugh, you complain about intptr-casted ints and then you got ahead and do this, have you heard of IntPtr.Zero?

Anyway Berkay, you don't understand what narcissism means, nowhere throughout our conversation I was being narcissistic and I don't care if you don't want to have IntPtr conversations and baseaddrs everywhere, until I will write a better memory class that doesn't have all of that mess and get rid of both VAMemory and GMemory.

**The bottom line is:** I don't have to respect your code if you don't respect the standards I set in NFSScript and then you got ahead and send me shit in PM like "suck my generic dick". You either stick to the code standards that NFSScript has or fuck right off.

**Comment by berkay2578** [comment]
umm.
https://github.com/DennisStanistan/NFSScript/blob/master/Core/VAMemory.cs#L2
http://i.imgur.com/sBHHzuM.png

ummm.
https://github.com/DennisStanistan/NFSScript/blob/master/Core/VAMemory.cs#L41

ummmm.
http://www.urbandictionary.com/define.php?term=joking%20around

**Comment by DennisStanistan** [comment]
Read the edited version.
